File: app/src/main/java/com/gameautopilot/app/brain/BrainResponseParser.kt
package com.gameautopilot.app.brain

import com.gameautopilot.app.core.Action
import com.gameautopilot.app.util.Logger
import org.json.JSONArray
import org.json.JSONObject

/**
 * Parses the strict-JSON `{thought, actions[], confidence, memory}` payload
 * that every Brain implementation asks the model for, once each provider has
 * unwrapped its own response envelope down to the raw model text.
 */
object BrainResponseParser {

    fun parse(rawModelText: String): BrainDecision {
        val cleaned = stripCodeFences(rawModelText).trim()
        val obj = try {
            JSONObject(cleaned)
        } catch (e: Throwable) {
            throw BrainException("Brain returned non-JSON: ${cleaned.take(200)}", e)
        }

        val thought = obj.optString("thought", "")
        val confidence = obj.optDouble("confidence", 0.5)
        val actionsArr = obj.optJSONArray("actions") ?: JSONArray()
        val actions = (0 until actionsArr.length()).mapNotNull { idx ->
            runCatching { Action.fromJson(actionsArr.getJSONObject(idx)) }
                .onFailure { Logger.w("Skipping bad action: ${it.message}") }
                .getOrNull()
        }
        val memoryUpdate = obj.optString("memory", "").ifBlank { null }
        return BrainDecision(thought = thought, actions = actions, confidence = confidence, memoryUpdate = memoryUpdate)
    }

    private fun stripCodeFences(s: String): String {
        val t = s.trim()
        if (!t.startsWith("```")) return t
        val firstNewline = t.indexOf('\n').takeIf { it >= 0 } ?: return t
        val withoutOpen = t.substring(firstNewline + 1)
        val endFence = withoutOpen.lastIndexOf("```")
        return if (endFence >= 0) withoutOpen.substring(0, endFence) else withoutOpen
    }
}

File: mac/Sources/GameAutopilotMac/Brain/BrainResponseParser.swift
import Foundation

/// Parses the strict-JSON `{thought, actions[], confidence, memory}`
/// payload every Brain implementation asks the model for, once each
/// provider has unwrapped its own response envelope down to raw model
/// text. Mirrors BrainResponseParser.kt.
enum BrainResponseParser {

    static func parse(_ rawModelText: String) throws -> BrainDecision {
        let cleaned = stripCodeFences(rawModelText).trimmingCharacters(in: .whitespacesAndNewlines)
        guard let data = cleaned.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw BrainError("Brain returned non-JSON: \(cleaned.prefix(200))")
        }

        let thought = (obj["thought"] as? String) ?? ""
        let confidence: Double
        if let d = obj["confidence"] as? Double {
            confidence = d
        } else if let n = obj["confidence"] as? NSNumber {
            confidence = n.doubleValue
        } else {
            confidence = 0.5
        }
        let actionsArr = (obj["actions"] as? [[String: Any]]) ?? []
        let actions = actionsArr.compactMap { Action.from(json: $0) }
        let memoryRaw = (obj["memory"] as? String) ?? ""
        let memoryUpdate = memoryRaw.isEmpty ? nil : memoryRaw
        let goalComplete = (obj["goalComplete"] as? Bool) ?? false

        return BrainDecision(
            thought: thought,
            actions: actions,
            confidence: confidence,
            memoryUpdate: memoryUpdate,
            goalComplete: goalComplete
        )
    }

    static func stripCodeFences(_ s: String) -> String {
        let t = s.trimmingCharacters(in: .whitespacesAndNewlines)
        guard t.hasPrefix("```") else { return t }
        guard let firstNewline = t.firstIndex(of: "\n") else { return t }
        let withoutOpen = String(t[t.index(after: firstNewline)...])
        if let endRange = withoutOpen.range(of: "```", options: .backwards) {
            return String(withoutOpen[withoutOpen.startIndex..<endRange.lowerBound])
        }
        return withoutOpen
    }
}
